Typed workflows (#6881)
nextflow-io/nextflow: docs/reference/operator-typed.md

The `cross` operator emits every pairwise combination of a source channel with another channel or value:

```{literalinclude} ../snippets/cross-v2.nf
:language: nextflow
```

```{literalinclude} ../snippets/cross-v2.out
:language: console
```

**`Channel<Record> cross( [opts] ) -> Channel<Record>`**

When the `cross` operator is called with named arguments and the source channel is a channel of records, the named arguments are appended to each record in the source channel. Each named argument can be a value or dataflow value.

## filter

**`Channel<E> filter( condition: (E) -> Boolean ) -> Channel<E>`**

The `filter` operator emits the values from a source channel that satisfy a condition, discarding all other values:

```{literalinclude} ../snippets/filter-closure.nf
:language: nextflow
```

```{literalinclude} ../snippets/filter-closure.out
:language: console
```

## flatMap

**`Channel<E> flatMap( transform: (E) -> Iterable<R> ) -> Channel<R>`**

The `flatMap` operator applies a *mapping function* to each value from a source channel. The mapping function should return a collection, and each element in the collection is emmited separately.

> For example:

Reviewer: ```suggestion
The `flatMap` operator applies a *mapping function* to each value from a source channel. The mapping function should return a collection, and each element in the collection is emmited separately:
```

:language: console
```

By default, unmatched values are discarded. The `remainder` option can be used to emit them at the end:

```{literalinclude} ../snippets/join-remainder-v2.nf
:language: nextflow
```

```{literalinclude} ../snippets/join-remainder-v2.out
:language: console
```

Available options:

`by: String`
: *This option is required.*
: The record field to use as the matching key.

`remainder: Boolean`
: When `true`, unmatched values are emitted at the end, otherwise they are discarded (default: `false`). 

## map

**`Channel<E> map( transform: (E) -> R ) -> Channel<R>`**

The `map` operator applies a *mapping function* to each value from a source channel:

```{literalinclude} ../snippets/map.nf
:language: nextflow
```

```{literalinclude} ../snippets/map.out
:language: console
```

## mix

**`Channel<E> mix( other: Channel<E> ) -> Channel<E>`**

**`Channel<E> mix( other: Value<E> ) -> Channel<E>`**

The `mix` operator emits the values from a channel and another channel or dataflow value into a single output channel:

```{literalinclude} ../snippets/mix.nf
:language: nextflow
```

```{literalinclude} ../snippets/mix.out
:language: console
```

> The values in the output channel may be emitted in any order, for example:

Reviewer: ```suggestion
The values in the output channel may be emitted in any order:
```

The `combine` operator emits every pairwise combination of a source channel with another channel or dataflow value:

```{literalinclude} ../snippets/combine.nf
:language: nextflow
```

```{literalinclude} ../snippets/combine.out
:language: console
```

Tuples in both the left- and right-hand sources are flattened in the combined tuple. For example, `tuple(1, 2)` and `tuple('red', 'blue')` are combined as `tuple(1, 2, 'red', 'blue')`.

**`Channel<Record> combine( [opts] ) -> Channel<Record>`**

When the `combine` operator is called with named arguments and the source channel is a channel of records, the named arguments are appended to each record in the source channel. Each named argument can be a value or dataflow value.

## filter

**`Channel<E> filter( condition: (E) -> Boolean ) -> Channel<E>`**

The `filter` operator emits the values from a source channel that satisfy a condition, discarding all other values:

```{literalinclude} ../snippets/filter-closure.nf
:language: nextflow
```

```{literalinclude} ../snippets/filter-closure.out
:language: console
```

## flatMap

**`Channel<E> flatMap( transform: (E) -> Iterable<R> ) -> Channel<R>`**

> The `flatMap` operator applies a *mapping function* to each value from a source channel. The mapping function should return a collection, and each element in the collection is emmited separately.

Reviewer: ```suggestion
The `flatMap` operator applies a *mapping function* to each value from a source channel. The mapping function should return a collection. Each element in the collection is emitted separately.
```
